designs/2025-define-plugin/README.md

```diff
@@ -0,0 +1,628 @@
+
+    ```js
+    export const plugin = {
+      configs: {
+        recommended: {
+          name: "example/recommended",
+          plugins: {
+            get example() {
+              return plugin;
+            },
+          },
+          rules,
+        },
+      },
+      meta: { name, version },
+      rules: recommendedRules,
+    };
+    ```
+
+  - Optimization requests such as [typescript-eslint/typescript-eslint#11029 Enhancement: Support Lazy Loading Rules](https://github.com/typescript-eslint/typescript-eslint/issues/11029) must be implemented per-plugin
+
+This RFC proposes creating a `definePlugin` function to be exported from `@eslint/plugin-kit`.
+It would provide a single recommended factory for defining the structure of a typical ESLint plugin.
+`definePlugin` would automate the parts of plugin definitions that many plugins manually redefine:
+
+- Using the same name in multiple places
+- Generating `configs.recommended` and other configurations based on rule metadata
+
+`definePlugin` would also allow plugins to ergonomically set up rules to be lazy-loaded.
+It would allow plugin authors to provide rules either as direct rule objects or as functions that return rules.
+An `eslint-plugin-eslint-plugin` lint rule would be created that keeps `configs` and `rules` in sync.
+
+As with `defineConfig`, `definePlugin` would be a purely optional, additive change.
+Existing and future plugins would continue to work as-is without a `definePlugin` call.
+
+## Detailed Design
+
+The `@eslint/plugin-kit` package would export a `definePlugin` function that takes in a single required options object.
+The options properties would require authors specify at least the following two properties:
+
+- `meta`: The same `name` as [Meta Data in Plugins](https://eslint.org/docs/latest/extend/plugins#meta-data-in-plugins), along with an optional `version`
+- `rules`: The same object as [Rules in Plugins](https://eslint.org/docs/latest/extend/plugins#rules-in-plugins), but also allowing functions that return a rule
+
+It would return a plugin with:
+
+- `configs`: by default, an object containing:
+  - `recommended`: a config array of a single object containing:
+    - `name`: per [configuration naming conventions](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-naming-conventions)
+    - `plugins`: an object keying `meta.name` to the plugin object
+    - `rules`: all rules whose `meta.docs?.recommended === true`, with severity set to `"error"`
+- all other provided properties as-is
+
+The following code block creates a standard plugin:
+
+```js
+import { definePlugin } from "@eslint/plugin-kit";
+
+import packageData from "../package.json" with { type: "json" };
+import { rules } from "./rules/index.js";
+
+export const plugin = definePlugin({
```

Comment:
A downside of this approach is that you can't tell just by looking at this code whether or not the plugin exports any configs.

Reply:
I like the idea of auto-generating and wiring up configs, I just feel like that we need to make that explicit in the API somehow. Off the top of my head, something like `autoConfig: true` would at least give a hint that something is happening behind the scenes.

Reply:
That makes sense to me. I'll go ahead and remove config automation from this RFC and leave `autoConfig: true` for a followup. I think the FC discussion has gone on long enough that we'd be best trying to land the core pieces without these nice-to-haves. 👍